### PR TITLE
Reformat sources with clang-format

### PR DESCRIPTION
**Pull request by reuk**
Uses clang-format 3.8 to reformat all sources using the default 'LLVM' style.

Clang-format has a few advantages over manual code formatting:
- Less developer effort spent thinking about whitespace when coding
- No developer effort wasted on formatting during code reviews
- Absolutely consistent (don't have to rely on reviewers catching all errors)
- No need for vague style guidelines written in English

Rebasing old patches can be achieved using a merge driver, similar to the method used when Doxygen-style comments were introduced. I've written a [wiki page](https://github.com/diffblue/doc-diffblue-common/wiki/Rebasing-PRs-after-the-introduction-of-clang-format) outlining the process. At present, I've left the modified `gitattributes` file in the commit history to facilitate faster rebases, as `develop` moves quite rapidly, and I want to keep this branch up-to-date. Of course, this commit should be dropped just before merging this PR.

On top of simply reformatting all the source, this PR will require some additional work before it is merged:
- Updating the linter to ignore old style rules
- Fixing regression tests so that they don't depend on sources being formatted in a particular style
- Adding some kind of style enforcement to CI
  - I would suggest to have a job which runs something like:
    ```bash
    clang-format -i **/*.(c|cpp|h|hpp)
    git diff > format.diff
    if [[ -s format.diff ]] ; then
        echo 'Formatting error!'
        echo 'Apply the following diff and resubmit'
        cat format.diff
        exit 1
    fi
    ```

Thoughts and feedback welcome!

### CODE REVIEW

**Comment by reuk** [comment]
Will solve https://diffblue.atlassian.net/projects/TG/issues/TG-451 for this repo.

**Comment by tautschnig** [comment]
I'll speak about why I think this is not a good idea below, but also I've just tried to scroll to one of the first files in src/ and got to see this:
```
-codet cpp_typecheckt::cpp_destructor(
-  const source_locationt &source_location,
-  const typet &type,
-  const exprt &object)
-{
+codet cpp_typecheckt::cpp_destructor(const source_locationt &source_location,
+                                     const typet &type, const exprt &object) {
```
If this is a serious suggestion, then at least the changes should be minimal rather than imposing a completely new coding style, and in particular one that hampers readability.

So why do I think this is not a good idea in the way this PR approaches this:
1. You do not necessarily know about the collection of patches that people maintain inside companies and outside any public repository. Re-formatting the source breaks all of them. Good bye customers.
2. For as long as code is written by humans, they should be trusted to apply sensible formatting, and in fact smart formatting: when I write code, I apply spacing and line breaks with good consideration. I break coding guidelines where it makes sense. My editor is configured to help me in this.
3. Everyone is free to run such a tool locally, e.g., when creating a new file.

Style guidelines are called "guidelines" for a reason, and they are kept in plain English because they apply to a particular audience. If we want to appeal to machines we should instead apply minification, it would make parsing and storage much more efficient.

**Comment by reuk** [comment]
> You do not necessarily know about the collection of patches that people maintain inside companies and outside any public repository. Re-formatting the source breaks all of them. Good bye customers.

I've addressed this on the diffblue-internal wiki page linked in my original post. The process for updating patches is similar to the process used when Doxygen was introduced: https://svn.cprover.org/wiki/doku.php?id=doxygen

Updating patches in this way is very straightforward. As long as your patches do not conflict with the commit before the reformatting change, the merge driver will automatically resolve all differences caused by formatting changes.

If customers are already used to rebasing their patches, the merge-driver approach adds very little overhead to the 'normal' git workflow (you just need to add a simple commit at the beginning of the commit range, and remove it again after the rebase). In fact, if they've pulled in the Doxygen update, they'll already know how to handle the clang-format change.

> For as long as code is written by humans, they should be trusted to apply sensible formatting, and in fact smart formatting: when I write code, I apply spacing and line breaks with good consideration. I break coding guidelines where it makes sense. My editor is configured to help me in this.

If we really trusted humans to apply sensible formatting, we wouldn't be using the linter script to catch style violations. Mostly, we want to keep the code consistent, though obviously the rules do not hold in all cases. For the linter, we manage special cases with `NOLINT` comments; for clang-format, there are similar comments (`// clang-format off` and `// clang-format on`) which can be used to disable auto-formatting where necessary. The main difference between the linter and clang-format is that the linter can't apply suggested changes automatically, protracting the development process.

> Everyone is free to run such a tool locally, e.g., when creating a new file.

At the moment this isn't possible. `clang-format` was designed to apply style guidelines from popular groups/projects (e.g. LLVM, Goole, Chromium, Mozilla, Webkit). The CBMC style is sufficiently different from other established projects that clang-format can't support it. For example:
- clang-format expects that spaces are always desirable around arithmetic operators
- clang-format expects that spaces around the inheritance `:` are desirable

This relates to your initial point ("at least the changes should be minimal rather than imposing a completely new coding style"), in that we *could* have a custom clang-format config which gets as-close-as-possible to the current style. But, why would we? Does the current style have concrete benefits over other possible styles? For this patch I used the LLVM style, but that's just because it's the default (and has been used successfully in LLVM and Clang).

> Style guidelines are called "guidelines" for a reason, and they are kept in plain English because they apply to a particular audience. If we want to appeal to machines we should instead apply minification, it would make parsing and storage much more efficient.

I agree with this up to a point. However, I find myself applying the same style rules in the same way a great deal when programming, and machines are *way better at this* than I am. Why not automate this process, and let the programmers worry about the important stuff like the actual semantics of the code they're writing (and picking good names for things).

**Comment by tautschnig** [comment]
Let me add another workflow that this approach breaks: the effective use of `git blame` to figure out which change may have introduced a particular line of code. This wholesale reformatting of the code base will mean that one has to distinguish whether a line as been introduced before or after reformatting.

Obviously this will only affect those that contribute by debugging issues.

**Comment by tautschnig** [comment]
Another day, another note: the develop/master set up and efforts of merging without altering commits would become impossible unless the reformatting is only put in place at a time where both branches are fully in sync.

**Comment by reuk** [comment]
> Let me add another workflow that this approach breaks: the effective use of git blame to figure out which change may have introduced a particular line of code.

That's true, and I don't have a good solution for this problem. However, looking at the diff, git doesn't seem to get confused about which "new" lines correspond to which "old" ones. Therefore, the git blame workflow would become:
- Run git blame
- Find offending line
- If the commit message for that line has "clang-format" in it, then run git blame again with the commit hash of the commit before the reformat (the blame interface on github makes this really easy)

**Comment by tautschnig** [comment]
Yes, this workaround workflow is what I would have applied, but it does remain a workaround and an extra complication.

My very personal summary just remains that I see no benefit in this proposed change, but instead a lot of extra complications. My view would likely be different if the changes were minimal as I do acknowledge the value of reducing the number of "change a whitespace here" comments on code reviews.

Alas, this is my view. It might be time for others with a different view to chime in.

**Comment by thk123** [comment]
# git blame

This is true, though there are tools designed to work around this ([git hyper blame](https://commondatastorage.googleapis.com/chrome-infra-docs/flat/depot_tools/docs/html/git-hyper-blame.html) allows you to ignore specific commits, e.g. the commit that re-formats everything - not tried this but figured there must be a way). I use git-blame very infrequently compared to reading and writing code, though maybe your uses-case is different. 

# `develop` vs. `master`

Was the intention to have `develop` be strictly ahead of `master`? If so - then we apply this to master, use the update process to this branch as a one off operation. If not, then I agree this is more of a problem, I suppose we would have to merge this into master first. 

# Value of the change

Of course for this inconvenience there must be some pay off. For me, it is all about an improved "quality of life" for developers who no longer have to worry about manual formatting. 

Everyone's first few PRs are littered with review comments pointing out spacing problems that waste valuable developer time spotting, fixing and rebasing. This means the content doesn't get a proper review and demoralises new developers (I am sure this is true non-DiffBlue contributors too). 

I'm not going to quibble over whether this format is the best, and if there is a strong preference for a different style I wouldn't be arguing for LLVM (personally hate the braces on same line for example :stuck_out_tongue:). However, the current CProver code style is not automatically formattable by any existing tools that I am aware of.

**Comment by owen-mc-diffblue** [comment]
I fully support the use of clang-format. For me, there are two main benefits. The first is that we will save a lot of time and interactions during code reviews. The second is that the style will really be consistent - currently different areas in the codebase have slightly different styles. I also welcome the fact that I will be able to set up a shortcut in my IDE to format my code correctly, so I don't have to think about what the formatting should be while I'm also trying to think about the problem at hand.

As for the style that is chosen, I think we should choose one of the standard ones, for simplicity. I haven't looked at them. The exact style we choose is less important to me than the consistency and simplicity of use.

This series of articles my MongoDB about moving over to clang-format is quite informative: https://engineering.mongodb.com/post/succeeding-with-clangformat-part-1-pitfalls-and-planning/

**Comment by tautschnig** [comment]
> This series of articles my MongoDB about moving over to clang-format is quite informative: https://engineering.mongodb.com/post/succeeding-with-clangformat-part-1-pitfalls-and-planning/

"Planning and rolling out the use of a formatting tool is not too hard; but it requires forethought, coordination, and a commitment to enabling and enforcing its use."

Maybe there are a few learnings that should be taken away from what's been going on here.

**Comment by owen-mc-diffblue** [comment]
"In short, employing an automated tool for code formatting has fantastic benefits. While the initial integration is expensive, it pays for itself in the long term in reduced costs writing and reviewing code."

The initial cost only gets larger the longer we leave it. I agree that we should take every care to make the transition as easy as possible, but we should do it sooner rather than later.

**Comment by tautschnig** [comment]
> The initial cost only gets larger the longer we leave it. I agree that we should take every care to make the transition as easy as possible, but we should do it sooner rather than later.

If people put the same effort into managing the develop -> master transition then I see no problem with it.

**Comment by reuk** [comment]
I had some free time so I took another look at this and tried to address some of the concerns expressed above.

- Added a clang-format config file which gets as close as possible to the previous style, and re-ran the formatting pass using the new style.
- Added a step to travis which checks that the formatting is correct, and fails the build if not. This could be made smarter by running only on changed files. However, it's already faster than the Python linter script so maybe that's not necessary.
- Adjusted the set of default filters in the linter to disable formatting-related checks. The argument could be made for removing the redundant code entirely.

**Comment by owen-mc-diffblue** [comment]
The attempt to get as close as possible to the existing style produces lines like
```changed= true;```
I don't like the way this looks. I think if we're going to have a space on one side of the `=` then we should have it on both sides.

**Comment by martin-cs** [comment]
Will break all existing patches and third party development.  Doesn't conform with existing coding style resulting in unnecessary, unimportant changes.  Does not resolve any issues nor add any feature.  IMHO do not merge.

**Comment by owen-mc-diffblue** [comment]
It temporarily makes things harder for third party developers and people with existing patches. The fact there's an automatic formatter to update their code means this isn't that hard to solve (cf. the MongoDB article I posted before). It makes future development easier, saves a lot of time in code review and makes life easier for people who want to start working on CBMC in future. It depends how you prioritise the past versus the future development of CBMC. I think the project will grow and grow, so doing this now will save a lot of time and effort in the long term.

**Comment by reuk** [comment]
I've fixed up this PR so that it
- Updates the linter to disable whitespace checks
- Adds a travis stage to run clang format, but *only over the diff*
- Updates the coding guidelines with information about the new formatting system

In this way, we can slowly move to a completely clang-formatted codebase, but without a single massive breaking change.

File: CODING_STANDARD.md [hunk 1]
```diff
@@ -1,46 +1,33 @@
...
-- Use `{}` instead of `;` for the empty statement
-- Single line blocks without `{ }` are allowed, but put braces around
-  multi-line blocks
-- Use blank lines to visually separate logically cohesive code blocks within a
-  function
-- Have a newline at the end of a file
+
+- Formatting and whitespaces are enforced using clang-format-3.8. The
+  formatting will be checked during the CI run, and style issues will be
+  displayed at this point.
```
Comment (tautschnig) [inline CODING_STANDARD.md]: Can we please retain some summary of the formatting rules as plain text? As is, it's required to reverse-engineer the clang-format configuration in order to know what to do. Yet, clang-format is the ultimate formalisation, but being stripped of any guideline is a rather poor experience.

File: scripts/cpplint.py [hunk 2]
```diff
@@ -224,8 +224,9 @@
     'readability/inheritance',
     'readability/multiline_comment',
     'readability/multiline_string',
-    'readability/namespace',
+    'readability/identifier_spacing',
```
Comment (tautschnig) [inline scripts/cpplint.py:227]: I don't think we should/need to skip these?

File: scripts/cpplint.py [hunk 3]
```diff
@@ -281,7 +282,12 @@
 # flag. By default all errors are on, so only add here categories that should be
 # off by default (i.e., categories that must be enabled by the --filter= flags).
 # All entries here should start with a '-' or '+', as in the --filter= flag.
-_DEFAULT_FILTERS = ['-build/include_alpha']
+_DEFAULT_FILTERS = [
+    '-build/include_alpha',
+    '-whitespace/',
+    '-readability/braces',
+    '-readability/identifier_spacing',
```
Comment (tautschnig) [inline scripts/cpplint.py]: None of the above change to the extent that cpplint would even notice, as far as I am aware. Notably, my understanding of the Friday meeting was that braces stay as-is.

File: using_clang_format.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
+# Using clang-format
```
Comment (tautschnig) [inline using_clang_format.md]: Don't introduce a new document that isn't linked from anywhere. Place this in "COMPILING" or at least add some links from that file to this one. In short: don't make it harder for people to contribute.

File: CODING_STANDARD.md (hunk 1 above)
Comment (reuk) [reply CODING_STANDARD.md]: I've left the rules as they were previously, with a couple of small alterations. Added a sentence referring to the clang-format info in COMPILING.md.

File: using_clang_format.md (hunk 4 above)
Comment (reuk) [reply using_clang_format.md]: Moved this into COMPILING.md

File: scripts/cpplint.py (hunk 3 above)
Comment (reuk) [reply scripts/cpplint.py]: At the same time, if clang-format is the source-of-truth for this, why bother verifying it with another tool?

Comment (tautschnig) [reply scripts/cpplint.py]: `cpplint` appears to be highly portable, and indeed very easy to set up. The same isn't necessarily true for `clang-format`. We might do this once we have built a lot of trust in `clang-format`. But why do it now?

File: scripts/cpplint.py (hunk 2 above)
Comment (reuk) [reply scripts/cpplint.py:227]: Identifier-spacing was added for the rule about `:` when inheriting. Previously, this rule was labeled `readability/naming`, but there are other useful rules with this label that we want to keep enabled.

There is no need to keep the `:` rule enabled, as it is now enforced by clang-format. There's no need to check it again here.

**Comment by reuk** [comment]
@tautschnig I've responded to your feedback

File: scripts/cpplint.py (hunk 2 above)
Comment (tautschnig) [reply scripts/cpplint.py:227]: See https://github.com/diffblue/cbmc/pull/1324#discussion_r144848784 -- I'd be fine with you disabling rules where `cpplint` now wrongly reports a mistake just because `clang-format` does it slightly differently. But I'd rather not have rules disabled just because we can auto-generate the correct code as I fail to see the value of doing so.

File: scripts/cpplint.py (hunk 3 above)
Comment (reuk) [reply scripts/cpplint.py]: This PR enables clang-format on Travis, so it doesn't *need* to be portable. (I mean, it *is* portable, and there's packages on most distributions for it.) The point is having a single source of truth for formatting rules. Having similar rules checked by two tools just allows for them to get out of sync, and at some point someone will end up asking whether to follow the linter or clang-format. Why not just make it clear from the outset that clang-format is the arbiter of style now?

File: scripts/cpplint.py (hunk 2 above)
Comment (reuk) [reply scripts/cpplint.py:227]: If we're saying "whatever clang-format does is correct" then I don't see the point in doing extra duplicated work in this script. The two outcomes are:
- the script gives the same results as clang-format (we did unnecessary work), or
- the script flags up a false positive and we defer to clang-format anyway (we also did unnecessary work

Besides, it's better separation of concerns to let clang-format do 'just style stuff' and to restrict this script to more semantic things.

File: scripts/cpplint.py (hunk 3 above)
Comment (tautschnig) [reply scripts/cpplint.py]: > This PR enables clang-format on Travis, so it doesn't need to be portable. (I mean, it is portable, and there's packages on most distributions for it.)

I can run `cpplint` on all my systems before pushing or even completing any commit. This isn't true of `clang-format`, unfortunately. (Yes, some of my systems don't have `clang-format` available.)

You do have a point in the split-brain problem, but I'd rather risk getting there than switching over without backup.

Comment (reuk) [reply scripts/cpplint.py]: OK, how about we re-enable these flags so that they are run by default locally, but disable them in Travis so that we don't ever get conflicting results from CI?

File: scripts/cpplint.py (hunk 2 above)
Comment (tautschnig) [reply scripts/cpplint.py:227]: You're highlighting how much each of us trust `clang-format`: you seem to fully trust it, whereas I'd rather not trust it just yet. Hence I'd happily add a third item "clang-format got it wrong and the script flags this."

I'll downgrade my "request for changes" to a comment and let someone else be the arbiter.

**Comment by reuk** [comment]
I've adjusted cpplint so that it still behaves like it used-to when run locally, but on Travis some of the whitespace checks are disabled so that clang-format can become the single source-of-truth for formatting.

File: scripts/run_diff.sh [hunk 5]
```diff
@@ -37,7 +37,7 @@ then
     echo "Ensure cpplint.py is inside the $script_folder directory then run again"
     exit 1
   else
-    cmd='${script_folder}/cpplint.py $file 2>&1 >/dev/null'
+    cmd='${script_folder}/cpplint.py $file --filter=-whitespace,-readability/identifier_spacing 2>&1 >/dev/null'
```
Comment (peterschrammel) [inline scripts/run_diff.sh]: Does this disable all whitespace checks? Didn't you want to disable only those for assignment operators?

Comment (reuk) [reply scripts/run_diff.sh]: Fixed.

File: scripts/cpplint.py (hunk 2 above)
Comment (martin-cs) [reply scripts/cpplint.py:227]: My feeling is that we should only disable the rules that clang-format *cannot* meet in any configuration.  Not everyone will use clang-format and we will still be using the linter on CI.  Hell; maybe we should add something to the end of the linting output so that if it finds errors it says, "why not use scripts/whatever.sh to fix this".

File: scripts/cpplint.py (hunk 3 above)
Comment (martin-cs) [reply scripts/cpplint.py]: Disable only the ones that clang-format would break; keep the rest.

File: COMPILING [hunk 6]
```diff
@@ -1,267 +0,0 @@
...
-To get coverage metrics run the following script from the regression
-directory:
-
-  get_coverage.sh
-
-This will:
- 1) Rebuild CBMC with gcov enabled
- 2) Run all the regression tests
- 3) Collate the coverage metrics
- 4) Provide an HTML report of the current coverage
```
Comment (martin-cs) [inline COMPILING:267]: Bundling a file rename along with a change to the content seems somewhat poor form as it makes it hard to see what is changed.

File: COMPILING.md [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+install clang-format locally and run it against your code as you are working.
+
+Different versions of clang-format have slightly different behaviors. CBMC uses
+clang-format-3.8 as it is available the repositories for Ubuntu 16.04 and
+Homebrew.
+To install, run:
+```
+sudo apt install clang-format-3.8 # Run this on Ubuntu, Debian etc.
+brew install clang-format-3.8     # Run this on a Mac with Homebrew installed
+```
```
Comment (martin-cs) [inline COMPILING.md]: We give build instructions for more platforms than just Ubuntu and Mac OS.  Any  chance of commands / pointers for those?

File: .travis.yml [hunk 8]
```diff
@@ -3,6 +3,32 @@ language: cpp
...
+        git diff --color > formatted.diff
+        if [[ -s formatted.diff ]] ; then
+          echo 'Formatting error! Apply the following diff and resubmit:'
+          cat formatted.diff
+          exit 1
+        fi
+        echo 'No formatting errors found'
+        exit 0
+      before_cache:
+
```
Comment (martin-cs) [inline .travis.yml:32]: I'm unsure as to whether this should be another commit or not.  In some ways it would be nice to roll the tool out locally, let people use it, then add it to CI.

File: COMPILING (hunk 6 above)
Comment (reuk) [reply COMPILING:267]: If I look at 1dcc82c github shows line-by-line changes, so I'm not sure this is a big issue.

File: .travis.yml (hunk 8 above)
Comment (reuk) [reply .travis.yml:32]: I don't really understand the drawbacks of enabling this straight away. Could you give some examples?

Comment (tautschnig) [reply .travis.yml:32]: It may wrongly block other PRs from passing CI.

File: COMPILING (hunk 6 above)
Comment (tautschnig) [reply COMPILING:267]: Weird that github does this when looking at a single commit, but doesn't get it when looking at all the commits in a PR.

File: .travis.yml (hunk 8 above)
Comment (reuk) [reply .travis.yml:32]: Unless there's a way of enabling travis checks just for new branches, I believe this will be an issue whenever we enable this check. In any case, the formatting can be corrected using the instructions in the COMPILING.md (the 'retroactively formatting individual commits' section).

File: COMPILING.md (hunk 7 above)
Comment (reuk) [reply COMPILING.md]: I've updated with instructions for more platforms

File: COMPILING.md [hunk 9]
```diff
@@ -236,3 +236,73 @@ This will:
...
+
+To avoid waiting until you've made a PR to find formatting issues, you can
+install clang-format locally and run it against your code as you are working.
+
+Different versions of clang-format have slightly different behaviors. CBMC uses
+clang-format-3.8 as it is available the repositories for Ubuntu 16.04 and
+Homebrew.
+To install, run:
+```
+sudo apt install clang-format-3.8 # Run this on Ubuntu, Debian etc.
```
Comment (tautschnig) [inline COMPILING.md]: `apt-get install` (missing "-get", remove "sudo" as that's not done elsewhere in this file)

File: COMPILING.md [hunk 10]
```diff
@@ -236,3 +236,73 @@ This will:
...
+brew install clang-format-3.8     # Run this on a Mac with Homebrew installed
```
Comment (tautschnig) [inline COMPILING.md]: In my `brew` set up I've got "clang-format" (seems to be version 4.0) or "clang-format@3.8". Is my set up wrong or is the above guideline wrong?

File: .travis.yml (hunk 8 above)
Comment (thk123) [reply .travis.yml:32]: @reuk would this step block a PR? The current linter, if fails, doesn't fail the PR. This shouldn't either, since there may be unforseen problems. 

However, providing it doesn't then I would have thought turning it on in this PR would be the quickest way to find out if there are issues (as opposed to relying on people doing it locally).

File: COMPILING.md [hunk 11]
```diff
@@ -236,3 +236,73 @@ This will:
...
+
+## RUNNING CLANG-FORMAT LOCALLY
+
+### INSTALLATION
+
+To avoid waiting until you've made a PR to find formatting issues, you can
+install clang-format locally and run it against your code as you are working.
+
+Different versions of clang-format have slightly different behaviors. CBMC uses
+clang-format-3.8 as it is available the repositories for Ubuntu 16.04 and
```
Comment (tautschnig) [inline COMPILING.md:270]: Do we need version 3.8 or could it be "any newer version"?

File: COMPILING.md (hunk 10 above)
Comment (reuk) [reply COMPILING.md]: I got the guideline wrong, fixed now

File: .travis.yml (hunk 8 above)
Comment (reuk) [reply .travis.yml:32]: At the moment it does block, but maybe if we made it non-blocking like the linter that would be a good solution - we can always make it blocking in the future.

File: COMPILING.md (hunk 11 above)
Comment (reuk) [reply COMPILING.md:270]: Subsequent versions have slightly different behaviour, so we need to standardise on a particular version.

File: .travis.yml (hunk 8 above)
Comment (reuk) [reply .travis.yml:32]: Done

File: COMPILING.md (hunk 11 above)
Comment (tautschnig) [reply COMPILING.md:270]: Is it just me that I find this lock-in scary?

Maybe I'm misunderstanding your comment: From my point of view this means that the premises 1) "Absolutely consistent (don't have to rely on reviewers catching all errors)" only holds when sticking with this one version forever and 2) "No need for vague style guidelines written in English" is actually untrue, because we'd still need some ground truth when forced to move to a newer version (clang-format 3.8 may become unavailable in pre-compiled form at some point).

Would you be able to provide more detail on "slightly different behaviour?"

Comment (reuk) [reply COMPILING.md:270]: As far as I can tell, newer versions add more formatting parameters, which may affect how clang-format chooses where to place line-breaks, but the config files are backward-compatible. That is, a newer version will produce a style that still obeys all the rules from the config file, but which may find better line-break positions (for some value of 'better').

I don't think that being locked to a specific version is too worrying, as we're already incrementally updating the style. If we want to update to a new version in the future, we can carry on incrementally updating, but using a newer clang-format binary.

Comment (tautschnig) [reply COMPILING.md:270]: I do feel better given the information you have just given me, but this isn't the message that the documentation is sending out right now. So why are you proposing specifically version 3.8 then? Is it because that's the version that is available in current distributions? If so, I'd suggest that package-manager install instructions refer to such a version (much like is done for GCC), but I would refrain from doing so in plain text.

What I want to avoid is a cliff edge when moving to a newer version. I'd rather move with whichever version is available in the distribution used by Travis. Your comments assure me that our incremental application of `clang-format` should not cause any breakage.

Comment (reuk) [reply COMPILING.md:270]: Yes, version 3.8 is available in packages for most platforms, so it seemed like a good choice. However, Ubuntu and Homebrew both supply newer versions of clang-format too, and typing `apt install clang-format` will default to the newer version. For this reason, I think it's important to provide install instructions specifically for version 3.8.

**Comment by reuk** [comment]
I've responded to all comments and CI is passing - is this ready to merge now?

File: .travis.yml [hunk 12]
```diff
@@ -165,7 +192,8 @@ jobs:
       script: echo "This is coverity build. No need for tests."
 
   allow_failures:
-    - <<: *linter-stage
+   - <<: *formatting-stage
+   - <<: *linter-stage
```
Comment (tautschnig) [inline .travis.yml]: Why has indentation changed here?

Comment (reuk) [reply .travis.yml]: That was an accident, fixed now

**Comment by reuk** [comment]
Passed CI again. Is everyone happy with the state of this now?

File: .travis.yml (hunk 8 above)
Comment (martin-cs) [reply .travis.yml:32]: Non-blocking is good.  It seems like the kind of thing that it might take a little while to get used to and to get to love, so making it blocking on day 1 seems a bit hasty.  There may also be unforseen things that we need to iron out and I'd rather not block commits on CI while that happens.

File: COMPILING (hunk 6 above)
Comment (martin-cs) [reply COMPILING:267]: OK; may just be a UI issue then.

File: COMPILING.md (hunk 11 above)
Comment (martin-cs) [reply COMPILING.md:270]: :-\  I find it a little concerning that they aren't offering settings that are backwards compatible.

**Comment by martin-cs** [comment]
Yeah; why not?  I'm in.
